### PR TITLE
[WIP] Fix vehicle-vehicle collisions

### LINKED ISSUE #28534: Colliding cars that face to different directions from each other will collide wrong
**Describe the bug**  
The velocities used in collisions are relative to the cars themselves. 
But the code assumes that the velocity vectors of both cars are in same coordinates.

Or something like that. I am not fully certain on the cause.

As a result cars that are facing to different directions will get wrong results when they collide.

**To Reproduce**  
You will need to add some extra messages in the code to see the vectors.

Something like this added to map.cpp at the end of `map::vehicle_vehicle_collisionfunction` will show it.
```
add_msg( m_info, _( "v_1:%fx + %fy ") , velo_veh1.x, velo_veh1.y  );
add_msg( m_info, _( "v_2:%fx + %fy ") , velo_veh2.x, velo_veh2.y  );
add_msg( m_info, _( "vcy_1 %f, vcy_2 %f") , vel1_y, vel2_y  );
```

Some situations where the problem can be seen during normal gameplay:
* Front ending a car does not push it and causes massive damage (collision tries to push the car to wrong direction causing more collisions to occur and each collision adds more velocity to wrong direction making each collision even more damaging). 
* The above can cause cars to get stuck when they colide at angle and the velocity is low enough to not destroy parts.
* Hitting a car from forward angle can push the car towards the collision.

**Expected behavior**  
Car The collided car should be pushed away from the collision, not towards it.
The result is wrong in various ways.

**Screenshots**  
Example 
Blue: Velocity before collision
Green: Expected velocity after collision
Red: The actual velocity after collision
a

**Versions and configuration(please complete the following information):**
 - OS: Windows 10
 - Game Version: 0.C-37605-g7a48668 (tiles)
 - Graphics version: Tiles

**Additional context**

### PR DESCRIPTION
**Pull request by Hirmuolio**
#### Summary

```SUMMARY: Bugfixes ""```  

#### Purpose of change
Tried to fix bug. Bugs not here.

#### Describe the solution
Currently the vehicle collision uses vehicle velocities and vehicle facing directions.
Vehicle collision was rewritten to use only vehicle velocities.

I don't know why the collision axis was calculated from vehicle orientation.
Now the collision is handled as if it happened along vehicle 1 velocity axis. Movement on other directions is not considered for the collision but it is added back after calculations are done.

#### Additional context
The collision code itself is mostly ready. Needs some working with how skidding works.

### CODE REVIEW

**Comment by Hirmuolio** [comment]
Anyone got idea how to change how skidding works?
Currently if you collide with a car it can only move forwards unless you hit the car hard enough to make it skid.
It would make sense that the car can move also backwards without needing to skid.

**Comment by Photoloss** [comment]
>  // There is no solution for the 2D case where vehicle 1 and vehicle 2 both are moving at some arbitrary angle

What exactly do you mean to say with this statement? This is a two-body problem and has well-defined analytical solutions. At worst you could always calculate the coordinate transformation into the co-moving frame of one vehicle (or the centre-of-mass frame) and deal with a single "effective" particle.

And while I doubt the game actually implements such a level of realism the "facing direction" of vehicles is crucial because their wheels can only roll along one axis. A sideways collision will immediately incur dragging friction which prevents recoiling from the impact much more strongly than rolling. The standard collision formula assumes both particles to be frictionless.

File: src/map.cpp [hunk 1]
```diff
@@ -602,59 +602,49 @@ float map::vehicle_vehicle_collision( vehicle &veh, vehicle &veh2,
     // Vertical collisions will be simpler for a while (1D)
     if( !vertical ) {
         // For reference, a cargo truck weighs ~25300, a bicycle 690,
-        //  and 38mph is 3800 'velocity'
-        rl_vec2d velo_veh1 = veh.velo_vec();
-        rl_vec2d velo_veh2 = veh2.velo_vec();
+        //  Velocities are in 1/100 mph
+        // Masses are in kilograms
+        rl_vec2d velocity_veh1 = veh.velo_vec();
```
Comment (mlangsdorf) [inline src/map.cpp]: while I am all for making variable names more descriptive, velo_veh1 was adequate and I feel this rename is unnecessary churn.

File: src/map.cpp [hunk 2]
```diff
@@ -602,59 +602,49 @@ float map::vehicle_vehicle_collision( vehicle &veh, vehicle &veh2,
...
+        rl_vec2d velocity_veh2 = veh2.velo_vec();
         const float m1 = to_kilogram( veh.total_mass() );
         const float m2 = to_kilogram( veh2.total_mass() );
         //Energy of vehicle1 and vehicle2 before collision
-        float E = 0.5 * m1 * velo_veh1.magnitude() * velo_veh1.magnitude() +
-                  0.5 * m2 * velo_veh2.magnitude() * velo_veh2.magnitude();
-
-        // Collision_axis
```
Comment (mlangsdorf) [inline src/map.cpp]: What happened to all this code?  Why did you remove it?

File: src/map.cpp [hunk 3]
```diff
@@ -602,59 +602,49 @@ float map::vehicle_vehicle_collision( vehicle &veh, vehicle &veh2,
...
-        veh.move.init( final1.x, final1.y );
-        veh.velocity = final1.magnitude();
-
-        veh2.move.init( final2.x, final2.y );
-        veh2.velocity = final2.magnitude();
-        //give veh2 the initiative to proceed next before veh1
+        // There is no solution for the 2D case where vehicle 1 and vehicle 2 both are moving at some arbitrary angle
+        // We simplify the situation to 1D case along the vehicle 1 velocity vector
+        // If vehicle 2 is stationary or moving along the same vector as vehicle 1 the result is 100% right
+        // If vehicle 2 is moving fast in a completely different direction the result may be a bit odd
```
Comment (mlangsdorf) [inline src/map.cpp]: Consider calculating two collisions: a 1D calculation along vehicle 1's velocity vector, and a 1D calculation on vehicle 2's velocity vector.

File: src/map.cpp [hunk 4]
```diff
@@ -602,59 +602,49 @@ float map::vehicle_vehicle_collision( vehicle &veh, vehicle &veh2,
...
+
+        const float component_velocity_veh1 = velocity_veh1.magnitude();
```
Comment (mlangsdorf) [inline src/map.cpp]:
```suggestion
        const float comp_velo_veh1 = velo_veh1.magnitude();
```
along with a note that this is the component of vehicle 1's velocity vector along the axis of vehicle 1's velocity vector.

These fairly long variable names become unwieldy immediately.  For instance, replacing:
```
        float vel1_y_a = ( m2 * vel2_y * ( 1 + e ) + vel1_y * ( m1 - m2 * e ) ) / ( m1 + m2 );
```
with
```
        const float component_velocity_veh1_after = ( elasticity * m2 * ( component_velocity_veh2 -
￼          		￼                 component_velocity_veh1 ) + m1 * component_velocity_veh1 + m2 * component_velocity_veh2 ) /
￼         		￼                 ( m1 + m2 );
```
does not really improve readability in my mind.

**Comment by Hirmuolio** [comment]
There probably is some solution for 2D case but quick google didn't give ready formula. And since practically all collisions in the game have one car moving and other being stationary (or being pushed) the 1D solution is good enough. Probably could have worded that better and will removed it now.

not-forwards movement needs some work still.

File: src/map.cpp (hunk 2 above)
Comment (Hirmuolio) [reply src/map.cpp]: It didn't go anywhere. Github just marked it removed due to small name change.

File: src/map.cpp (hunk 3 above)
Comment (Hirmuolio) [reply src/map.cpp]: I thought about that but decided that the extra work for proper 2d math is not worth it.
I can't think of any realistic situation where you would collide with a car that moves at a considerable sideways velocity compared to your own car.

All cars that you push move approximately same direction as your car and rest are stationary.

**Comment by Photoloss** [comment]
> There probably is some solution for 2D case but quick google didn't give ready formula.

See [this one](https://en.wikipedia.org/wiki/Elastic_collision#Two-dimensional) for the perfectly elastic case and [here](https://en.wikipedia.org/wiki/Impact_parameter#Scattering_from_a_hard_sphere) for how to substitute the scattering angle with something we can approximate within tile geometry. 

If you want to add additional complications like inelasticity or anisotropy (i.e. rolling wheels) those can be solved for as well but you first need to provide a formula for what is actually input from these models. For example the current *get_collision_factor* function is a simple linear model using only the difference between the initial velocities, with no indication as to how or why this was implemented. The true definition of [elasticity](http://physics.bu.edu/~redner/211-sp06/class-momentum/elasticity.html) includes the final state velocities which would make a proper solution far more difficult.

**Comment by mlangsdorf** [comment]
You should do a longer google search and find the 2D case.  Or follow @Photoloss' links; thank you very much for those, Photoloss.

**Comment by Hirmuolio** [comment]
Damnit now I accidentally closed this.

I'll try to implement the 2d calculation by calculating the two separate directions separately and reopen this tomorrow.

**Comment by Hirmuolio** [comment]
The 2D collision is now calculated as two separate 1 D collisions. One along the vehicle 1 velocity vector and another orthogonal to vehicle 1 velocity vector.

**Comment by Hirmuolio** [comment]
After further testing it seems that these changes were for nothing.

The problem isn't how collisions are calculated but how pushed vehicles move.

If a vehicle is pushed in a direction other than the direction it is facing it will do strange things.

At low velocities (slow enough that the vehicle doesn't actually move) the pushed vehicle seems to forwads with all of its velocity even if the velocity was backwards. I thought the collision was calculated in wrong direction but it isn't.

Stationary cars coliding will collide twice (first A collides with B and then B collides with A). The above problem causes them to push together again after collision causing the second collision and some strange thing about turns causes the two collisions to happen on same turn.

I'll leave this pull request as is for now as changing how the collisions are calculated will not fix the problem. If you think the old collision calculations should be replaced with something like this just say so.
The main difference is that the old code took some strange vector from the orientation of the two cars and calculated the collision only along this vector while my code calculates the collision along the car 1 velocity vector. As a result this code causes more damage (full speed is used instead of a partial component). Also 2d calculations but those won't matter in practice.

The two codes give in some cases identical results but usually similar looking but different result.

**Comment by Photoloss** [comment]
> If a vehicle is pushed in a direction other than the direction it is facing it will do strange things.

You can say that again, my motorised 1-tile cart tends to "collide with errors in the movement code" if I don't grab it from its "forwards" direction.

I can try and work out the physics of both codes but how large are the differences? Are we talking about a few % which could be due to rounding/int casting somewhere or are the values off by half? Coordinate transformations can look very weird if you aren't used to them so most people won't be able to tell the difference between a genius code optimisation and a terrible approximation. If you understand the basics of the physics you can plug in some simple edge cases (e.g. either vehicle at rest, equal masses, 1D in-line collision) and see if the coded formula produces the same simplified result as found in any textbook/wiki. Hopefully I'll have time for that check over the weekend.

**Comment by Hirmuolio** [comment]
The current code takes the vehicle orientation and relative positions and uses those to calculate a vector (I am not sure why it is done this way or what the vector is supposed to represent).

Then the collision is calculated with the velocities along this vector. Velocities in other direction are ignored in calculations and don't change.

The difference depends on velocity and angle. At certain angles the results are identical and in some angles the old code barely registered a hit compared to the new code.

Especially in situations where the player car steers on a car that is on their side the old code got very weak collision (sometimes literally zero value collision).

Since this code makes more big collisions the collision damage should be also reduced.

I uploaded more experimental version of the code. It has the new code doing all the work but it also calculates the result of the old code. It will then print the velocities in the game log in this form:
Vehicle 1 velocity vector before impact
Vehicle 2 velocity vector before impact

Vehicle 1 velocity vector after impact (new code)
Vehicle 2 velocity vector after impact (new code)
Vehicle 1 velocity vector after impact (old code)
Vehicle 2 velocity vector after impact (old code)

you can crash around and see how different the values are.

**Comment by mlangsdorf** [comment]
unit test cases go in tests/, not src/map.cpp.

**Comment by Hirmuolio** [comment]
Ok. I think I understand now how it works now (and why the new code I have is not good. I basically handled the vehicles as point masses and that isn't too good).

The current system gets vector from car 1 center of mass to car 2 center of mass. It just does it with some odd steps so I didn't understand it at first. it could be done in a single vector addition instead of assigning extra variables for the x- and y-components and calculating the vector components separatelely in four parts.

Then it takes the velocity vector component along this vector and uses this component for the collision. The other component of velocioty is ignored.

Black vector is from car center of mass to other car center of mass
Green is velocity vector
Red is component of velocity vector used for collision
Purple is velocity vector not used


I think the use of this component for collision is pretty good. But not using the other component for anything is a bit of a problem. For example the second image in above example would have high speed impact that would do basically nothing.

I was thinking that the other veloity vector component could apply some friction-like energy transfer between the vehicles. This would push the other car forwards when the collision happens in a very shallow angle.
